File: ecommerce-api/src/main/java/com/josue/ecommerce/cart/service/CartService.java
package com.josue.ecommerce.cart.service;

import com.josue.ecommerce.cart.dto.CartResponse;

import java.util.UUID;

public interface CartService {
    CartResponse createOrGetActive();
    CartResponse getCart(UUID cartId);
    CartResponse setQuantity(UUID cartId, UUID productId, int quantity);
    CartResponse removeProductFromCart(UUID cartId, UUID productId);
}

File: ecommerce-client/src/services/cartService.ts
import type { Cart, Order } from '../types'
import { jsonRequest, request } from './httpClient'

export const cartService = {
  createOrGetCart: async (): Promise<Cart> =>
    await request<Cart>('/api/v1/carts', { method: 'POST' }),

  getCart: async (cartId: string): Promise<Cart> => await request<Cart>(`/api/v1/carts/${cartId}`),

  setQuantity: async (cartId: string, productId: string, quantity: number): Promise<Cart> =>
    await request<Cart>(
      `/api/v1/carts/${cartId}/items/${productId}`,
      jsonRequest('PUT', { quantity }),
    ),

  removeItem: async (cartId: string, productId: string): Promise<Cart> =>
    await request<Cart>(`/api/v1/carts/${cartId}/items/${productId}`, {
      method: 'DELETE',
    }),

  checkout: async (cartId: string, idempotencyKey: string): Promise<Order> =>
    await request<Order>(`/api/v1/carts/${cartId}/checkout`, {
      method: 'POST',
      headers: { 'Idempotency-Key': idempotencyKey },
    }),
}
